### PR TITLE
curve-fuse: use bs aio interfance

### PR DESCRIPTION
**Pull request by wu-hanqing**
### What problem does this PR solve?

Issue Number: #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by wu-hanqing** [comment]
recheck

**Comment by wu-hanqing** [comment]
recheck

**Comment by wu-hanqing** [comment]
recheck

**Comment by wu-hanqing** [comment]
recheck

**Comment by wu-hanqing** [comment]
recheck

**Comment by wu-hanqing** [comment]
recheck

File: curvefs/src/volume/block_device_aio.cpp [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        aio.op = LIBCURVE_OP_READ;
+        aio.cb = AioReadCallBack;
+        aio.offset = offset;
+        aio.length = length;
+        aio.buf = data;
+
+        int ret = dev->AioRead(fd, &aio);
+        if (ret < 0) {
+            LOG(ERROR) << "Failed to issue aio read: " << &aio;
+            done = true;
```
Comment (cw123) [inline curvefs/src/volume/block_device_aio.cpp:111]: need atomic ?

File: curvefs/src/volume/block_device_aio.h [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    std::condition_variable cond;
+
+    struct Padding {
+        off_t offset;
+        size_t length;
+        std::unique_ptr<char[]> data;
+    };
+
+    // padding read request if necessary
+    std::unique_ptr<Padding> padding;
```
Comment (cw123) [inline curvefs/src/volume/block_device_aio.h:69]: need init nullptr when init？

File: curvefs/src/volume/block_device_aio.cpp [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+    if (static_cast<ssize_t>(aio.ret) != static_cast<ssize_t>(aio.length)) {
+        LOG(ERROR) << "AioRead error: " << &aio;
+        return -1;
+    }
+
+    if (padding != nullptr) {
+        std::memcpy(data, padding->data.get() + (offset - padding->offset),
+                    length);
+    }
```
Comment (cw123) [inline curvefs/src/volume/block_device_aio.cpp:152]: padding->data.reset(new char[padding->length]);

not free

File: curvefs/src/volume/block_device_aio.cpp [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+    aux = absl::make_unique<PaddingAux>();
+
+    aux->alignedOffset = align_down(offset, IO_ALIGNED_BLOCK_SIZE);
+    auto alignedEnd = align_up(offset + length, IO_ALIGNED_BLOCK_SIZE);
+
+    aux->shift = offset - aux->alignedOffset;
+    aux->alignedLength = alignedEnd - aux->alignedOffset;
+
+    aux->paddingData.reset(new char[aux->alignedLength]);
```
Comment (cw123) [inline curvefs/src/volume/block_device_aio.cpp:191]: where free this char[] ?

File: curvefs/src/volume/block_device_aio.cpp (hunk 1 above)
Comment (wu-hanqing) [reply curvefs/src/volume/block_device_aio.cpp:111]: in this case, it means we failed to start an async read request, and we mark this request as failed, so, there are no other threads that will notify it, only `Wait` will read it and is also called in the same thread.

File: curvefs/src/volume/block_device_aio.h (hunk 2 above)
Comment (wu-hanqing) [reply curvefs/src/volume/block_device_aio.h:69]: it's nullptr by default, the default constructor guarantee this.

File: curvefs/src/volume/block_device_aio.cpp (hunk 4 above)
Comment (wu-hanqing) [reply curvefs/src/volume/block_device_aio.cpp:191]: the unique_ptr's destructor will free the memory

File: curvefs/src/volume/block_device_aio.cpp (hunk 3 above)
Comment (wu-hanqing) [reply curvefs/src/volume/block_device_aio.cpp:152]: the unique_ptr's destructor will free the memory

File: curvefs/src/volume/block_device_client.cpp [hunk 5]
```diff
@@ -211,72 +181,50 @@ ssize_t BlockDeviceClientImpl::Write(const char* buf,
...
+    return error ? -1 : total;
 }
 
 bool BlockDeviceClientImpl::WritePadding(char* writeBuffer,
                                          off_t writeStart,
                                          off_t writeEnd,
-                                         off_t offset,
-                                         size_t length) {
+                                         off_t offset,     // actual offset
+                                         size_t length) {  // actual offset
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply curvefs/src/volume/block_device_client.cpp:227]: fix

File: curvefs/src/volume/block_device_client.cpp [hunk 6]
```diff
@@ -164,41 +145,30 @@ ssize_t BlockDeviceClientImpl::Readv(const std::vector<ReadPart>& iov) {
...
-
-        taskpool_.Enqueue(std::move(task));
+    std::vector<std::unique_ptr<AioRead>> requests;
+    requests.reserve(iov.size());
+
+    for (const auto& io : iov) {
+        requests.push_back(absl::make_unique<AioRead>(
+            io.offset, io.length, io.data, fileClient_.get(), fd_));
+
+        requests.back()->Issue();
```
[the remark this answers is not in the document]
Comment (wu-hanqing) [reply curvefs/src/volume/block_device_client.cpp:155]: `back()` is a reference to the last `AioRead` that was inserted in line 152, so the logic here is 1) creating a read request and 2) issuing it, then 3) saving it into a vector, and the code is writing in order of 1/3/2.

File: curvefs/test/volume/block_device_client_test.cpp [hunk 7]
```diff
@@ -325,33 +412,36 @@ TEST_F(BlockDeviceClientTest, TestWriteWithUnAligned) {
...
             .Times(0);
         ASSERT_LT(client_->Write(buf, 1000, 5000), 0);
 
         // CASE 3: read failed once -> write failed
-        EXPECT_CALL(*fileClient_, Read(10, _, 8192, 4096))
-            .WillOnce(Return(4096));
-        EXPECT_CALL(*fileClient_, Read(10, _, 16384, 4096))
-            .WillOnce(Return(4095));
-        EXPECT_CALL(*fileClient_, Write(_, _, _, _))
+        // EXPECT_CALL(*fileClient_, Read(10, _, 8192, 4096))
```
Comment (Wine93) [inline curvefs/test/volume/block_device_client_test.cpp:430]: why comment it?

**Review by Wine93 — APPROVED** [review]
LGTM!

**Review by cw123 — APPROVED** [review]

File: curvefs/test/volume/block_device_client_test.cpp (hunk 7 above)
Comment (wu-hanqing) [reply curvefs/test/volume/block_device_client_test.cpp]: removed, we use asynchronous interface now.

**Pull request merged**